src/rgw/rgw_rados.cc

```diff
@@ -12052,3 +12052,45 @@ librados::Rados* RGWRados::get_rados_handle()
   }
 }
 
+int RGWRados::delete_obj_aio(rgw_obj& obj, rgw_bucket& bucket,
+                              RGWBucketInfo& bucket_info, RGWObjState *astate,
+                              list<librados::AioCompletion *>& handles)
+{
+  rgw_rados_ref ref;
+  int ret = get_obj_ref(obj, &ref, &bucket);
+  if (ret < 0) {
+    lderr(cct) << "ERROR: failed to get obj ref with ret=" << ret << dendl;
+    return ret;
+  }
+
+  RGWRados::Bucket bop(this, bucket_info);
+  RGWRados::Bucket::UpdateIndex index_op(&bop, obj, astate);
+
+  ObjectWriteOperation op;
+  ret = index_op.prepare(CLS_RGW_OP_DEL);
+  if (ret < 0) {
+    lderr(cct) << "ERROR: failed to prepare index op with ret=" << ret << dendl;
+    return ret;
+  }
+
+  list<string> prefixes;
+  cls_rgw_remove_obj(op, prefixes);
+
+  AioCompletion *c = librados::Rados::aio_create_completion(NULL, NULL, NULL);
+  ret = ref.ioctx.aio_operate(ref.oid, c, &op);
+  if (ret < 0) {
+    lderr(cct) << "ERROR: AioOperate failed with ret=" << ret << dendl;
+    return ret;
+  }
+
+  handles.push_back(c);
+
+  ret = delete_obj_index(obj);
```

Comment:
@prallabh you're only removing the bucket index entry ... not the object itself.